### PR TITLE
Upstream some list functions

### PR DESCRIPTION
**Pull request by Geometer1729**
This PR includes some functions for dealing with lists and property tests for those functions.

### CODE REVIEW

File: Plutarch/List.hs [hunk 1]
```diff
@@ -363,3 +370,134 @@ plistEquals =
...
+      (pcon PNothing)
+      xs
+
+psort :: POrd a => Term s (PList a :--> PList a)
+psort = timSort
+
+-- TODO: decide on default sort
+-- probably by benchmarking them
+
+mergesort :: POrd a => Term s (PList a :--> PList a)
```
Comment (L-as) [inline Plutarch/List.hs:413]: Naming

File: Plutarch/List.hs [hunk 2]
```diff
@@ -363,3 +370,134 @@ plistEquals =
...
+              (pcons # y #$ self # xs # ys')
+
+-- most of the tricks from timsort are not implemented here
+-- just the central strategy of finding runs of
+-- consecutive data and then merging them together
+timSort :: POrd a => Term s (PList a :--> PList a)
+timSort = phoistAcyclic $
+  plam $ \xs -> merge2 #$ timSplit # xs
+
+timSplit :: POrd a => Term s (PList a :--> PList (PList a))
```
Comment (L-as) [inline Plutarch/List.hs:457]: Naming here too

File: plutarch.cabal [hunk 3]
```diff
@@ -167,19 +167,26 @@ test-suite examples
...
+    Properties.Utils
     Utils
 
   build-depends:
     , aeson
     , base
     , bytestring
     , generics-sop
+    , hedgehog
+    , mtl
```
Comment (L-as) [inline plutarch.cabal:182]: Why mtl?

**Review by L-as — CHANGES_REQUESTED** [review]
@srid What do you think about the testing?

File: Plutarch/List.hs (hunk 1 above)
Comment (Geometer1729) [reply Plutarch/List.hs]: Do you want all functions name psomeFunctionName or just exposed ones?

**Comment by srid** [comment]
Just spoke with Brian. The haskPlutEquiv stuff is valuable for property testing Plutarch functions by piggybacking on their Haskell equivalent functions (it avoids having to write manual tests for each of them). 

I feel [Properties/Utils.hs](https://github.com/Plutonomicon/plutarch/pull/274/files#diff-f8ff3e451db7fa0415b03a4d0d75303fc0234aa9d0296b4c42def5519a3b0403)'s export interface could be simplified though. But it is probably not important right now, as all of these testing code will have to moved to `./plutarch-test` (and thereby refactored) anyway after #276 is merged.

File: Plutarch/List.hs (hunk 1 above)
Comment (L-as) [reply Plutarch/List.hs]: Any export that is a Plutarch function or type from any module should be prefixed with `p`.

File: plutarch-test/src/Util.hs [hunk 4]
```diff
@@ -0,0 +1,286 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Util (
```
Comment (srid) [inline plutarch-test/src/Util.hs:5]: You want to move this to `Plutarch.Test.Property` inasmuch as the whole thing is to do with property testing.

File: plutarch-test/src/Gen.hs [hunk 5]
```diff
@@ -0,0 +1,54 @@
+module Gen (
```
Comment (srid) [inline plutarch-test/src/Gen.hs:1]: Should be `Plutarch.Test.Property.Gen`

File: plutarch-test/src/Plutarch/ListPropSpec.hs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+import Plutarch.Test (PlutarchGolden (All), goldens)
+import Test.Syd (Spec, describe, it)
+import Test.Syd.Hedgehog ()
+
+spec :: Spec
+spec = describe "list spec" $ do
```
Comment (srid) [inline plutarch-test/src/Plutarch/ListPropSpec.hs:21]: By convention we don't have spaces in spec descriptions that get used as golden filenames. This should be therefore `describe "list"`.

**Review by srid — CHANGES_REQUESTED** [review]

**Comment by Geometer1729** [comment]
Moved to #304 so it can be from a different branch.

**Pull request closed without merging**